LoopUI/Charts/DoseChart.swift

```diff
@@ -0,0 +1,176 @@
+
+        let doseArea = ChartPointsFillsLayer(
+            xAxis: xAxisLayer.axis,
+            yAxis: yAxisLayer.axis,
+            fills: [ChartPointsFill(
+                chartPoints: basalDosePoints,
+                fillColor: colors.doseTint.withAlphaComponent(0.5),
+                createContainerPoints: false
+            )]
+        )
+
+        let bolusLayer: ChartPointsScatterDownTrianglesLayer<ChartPoint>?
+
+        if bolusDosePoints.count > 0 {
+            bolusLayer = ChartPointsScatterDownTrianglesLayer(xAxis: xAxisLayer.axis, yAxis: yAxisLayer.axis, chartPoints: bolusDosePoints, displayDelay: 0, itemSize: CGSize(width: 12, height: 12), itemFillColor: colors.doseTint)
+        } else {
+            bolusLayer = nil
+        }
+
+        // Grid lines
+        let gridLayer = ChartGuideLinesForValuesLayer(xAxis: xAxisLayer.axis, yAxis: yAxisLayer.axis, settings: guideLinesLayerSettings, axisValuesX: Array(xAxisValues.dropFirst().dropLast()), axisValuesY: yAxisValues)
+
+        // 0-line
+        let dummyZeroChartPoint = ChartPoint(x: ChartAxisValueDouble(0), y: ChartAxisValueDouble(0))
+        let zeroGuidelineLayer = ChartPointsViewsLayer(xAxis: xAxisLayer.axis, yAxis: yAxisLayer.axis, chartPoints: [dummyZeroChartPoint], viewGenerator: {(chartPointModel, layer, chart) -> UIView? in
+            let width: CGFloat = 1
+            let viewFrame = CGRect(x: chart.contentView.bounds.minX, y: chartPointModel.screenLoc.y - width / 2, width: chart.contentView.bounds.size.width, height: width)
+
+            let v = UIView(frame: viewFrame)
+            v.layer.backgroundColor = colors.doseTint.cgColor
+            return v
+        })
+
+        if gestureRecognizer != nil {
+            doseChartCache = ChartPointsTouchHighlightLayerViewCache(
+                xAxisLayer: xAxisLayer,
+                yAxisLayer: yAxisLayer,
+                axisLabelSettings: axisLabelSettings,
+                chartPoints: allDosePoints,
+                tintColor: colors.doseTint,
+                gestureRecognizer: gestureRecognizer
+            )
+        }
+
+        let layers: [ChartLayer?] = [
+            gridLayer,
+            xAxisLayer,
+            yAxisLayer,
+            zeroGuidelineLayer,
+            doseChartCache?.highlightLayer,
+            doseArea,
+            doseLine,
+            bolusLayer
+        ]
+
+        return Chart(frame: frame, innerFrame: innerFrame, settings: chartSettings, layers: layers.compactMap { $0 })
+    }
+}
+
+public extension DoseChart {
+    func setDoseEntries(_ doseEntries: [DoseEntry]) {
+        let dateFormatter = DateFormatter(timeStyle: .short)
+        let doseFormatter = NumberFormatter.dose
+
+        var basalDosePoints = [ChartPoint]()
+        var bolusDosePoints = [ChartPoint]()
+        var allDosePoints = [ChartPoint]()
+
+        for entry in doseEntries {
+            let time = entry.endDate.timeIntervalSince(entry.startDate)
```

Comment:
Might make sense a computed variable on DoseEntry, something like `deliveryDuration`

Reply:
Agree, this will have to be a TODO for LoopKit.